### DISCUSSION TITLE
Roadmap to Production-Readiness

### DISCUSSION THREAD

**Discussion by chickeaterbanana** [microsoft/terraform-provider-power-platform#365] [General]
Hey,

I really like the progress, which is made here. I've the challenge to administer many different power platform environments and would love to automate this. This also would simplify the process of introducing new features or changes into all existing environments, without clicking through all of those.

This is the reason I'm searching for a good IaC solution for Power Platform. I used terraform in the past a lot and really liked it. But unfortunately you are explicit mentioning, this project is not intended to be used in production. Is there a roadmap of features, which are needed until you would classify it as production ready?

Are there any production ready solution available, which can be seamlessly be migrated to this terraform template after the production readiness is reached? related to: https://github.com/microsoft/terraform-provider-power-platform/issues/42 

Thanks for your time,

chickeaterbanana

**Comment by mawasile** [comment]
Hi

The project is marked as experiments as we still working on it. We don't have specific set of functionalities to mark it as "production ready" but we are working right now in a form of a private preview to polish things that we already have and find any big gaps in features. 

We have a plan to remove "experimental / preview" warning in next 3-6 months. 

As for the import of an existing tenants, we're not there yet but this is something we would like to have once marking as production ready. 
We are also working on terraform modules that will encapsulate common administrative tasks and we are planning to push that into terraform registry. 
We are hoping that all of this should be finished till end of calendar year.

**Comment by mattdot** [comment]
Adding to what @mawasile said.  One of our signals for when it's right to remove the "experimental" warning is feedback from people like yourself who have evaluated the provider and think that it is ready for your production scenarios.  I'd definitely encourage you to try out the provider for some of your use cases and give us feedback / file bugs for any issues that are blockers for you.

**Comment by webstean** [comment]
To me, this product is ready to go production when it can demonstrate it can seamlessly setup a best practice Power App environment, consisting of:

- [ ] Environment Groups with customisable Environment Group settings  
- [ ] Environments with customisable Environment settings
- [ ] Tenant settings
- [ ] S2S users, linked to application registrations with customisable roles
- [ ] Teams, linked to Entra ID groups, with customisable roles.
- [ ] DLP policies
- [ ] VNet integration
- [ ] Installation of a basic set of solutions for environments
- [ ] Setup of GitHub/ADO integration for either individual environments and/or solution (pending some Wave releases)

**Comment by webstean** [comment]
Perhaps a good measure of readiness, would be the provider has a single configuration example that demonstrates the installation of the [Power Platform CoE Starter Kit](https://learn.microsoft.com/en-us/power-platform/guidance/coe/overview) including customisation of the ProductionEnvironment environment variable and installation of dependencies like the [Creator Kit](https://learn.microsoft.com/en-us/power-platform/guidance/creator-kit/setup)

**Reply by mawasile** [reply]
We are already working on that. 
https://github.com/microsoft/power-platform-terraform-quickstarts/tree/mawasile/coe-kit-iac-quickstart

**Comment by webstean** [comment]
FYI: COE Starter Kit cannot be installed with a Service Principle, [Service Principles Limitations](https://learn.microsoft.com/en-us/power-platform/guidance/coe/limitations#service-principles)

**Reply by mattdot** [reply]
It's actually a little more nuanced.  Service Principal can create the connections it just can't authenticate for a connection that requires a user.  You can use SP to create so that you can wire up the connection to connection references and then have a user manually authenticate the connection afterwards.
